Add: Ability to use creation patterns for other post types besides page

jorgefilipecosta opened the pull request:
Closes: https://github.com/WordPress/gutenberg/issues/41304


This PR is a follow-up to https://github.com/WordPress/gutenberg/pull/40034 and allows the creation patterns modal to appear in other post types besides pages.

## API

In order for a pattern to appear in a post type starting modal, the pattern blockTypes need to include support for 'core/post-content' ( so it defines the pattern is supposed to be the full content of a post) and the pattern should define the postTypes it supports as starting content in the postTypes array. Defining support only 'core/post-content' continues to work as before as by default we assume that a pattern supports being the full content of a page.

```
register_block_pattern( 'custom-pattern', array(
			'title'      => _x( 'Start post pattern', 'Block pattern title', 'gutenberg' ),
			'blockTypes' => array( 'core/paragraph', 'core/post-content' ),
			'postTypes' = array( 'product' ),
			'content'    => '<!-- wp:paragraph -->
<p>A start post pattern</p>
<!-- /wp:paragraph -->',
		) );
```

## API shape alternatives
*Old discussion*:
I'm not totally sure if using blockTypes with  'core/post-content/$postType' is the right path. For template-parts we use a similar format e.g: 'core/template-part/header'.  This format is consistent with the template-pasts format and that is the main reason I'm using it.
If people prefer I'm open to an alternative format e.g: using a  new property on the pattern like "creationPostTypes" an array of post types that the pattern defines support for appearing as start content (e.g: ["page, "product",...]).

cc: @mtias, @mcsf in case you have a preference for a specific API shape.

Updated: following @mtias suggestion we use a post types array.

## Post types defining patterns they support vs Patterns defining the post types they support

In https://github.com/WordPress/gutenberg/issues/41304, @Humanify-nl proposed a different API, where the post type is the one that defines patterns that appear. I'm following a different path where the patterns define the post types where they appear. This allows patterns in the pattern directory to define support for the common post types where they work, without the post type not even knowing that the pattern exists. It seems post types are less numerous than patterns so it is more expected that the pattern creator knows the post types where the pattern makes sense than the post type creator is aware of all the possible patterns that can be used.



## Common use cases and code samples of how to achieve them.


### On my "Special Page" post type I want to support all the patterns that appear on normal pages

```
	$patterns = WP_Block_Patterns_Registry::get_instance()->get_all_registered();
	foreach ( $patterns as $pattern ) {
		if (
			! empty($pattern['blockTypes'] ) && in_array('core/post-content', $pattern['blockTypes'] ) &&
			( empty($pattern['postTypes'] || ( in_array('page', $pattern['postTypes'] ) && ! in_array('page', $pattern['special-page'] ) )
		) {
			unregister_block_pattern( $pattern['name'] );
			$pattern['postTypes'] = array_merge( empty($pattern['postTypes']) ? array() : $pattern['startContentPostTypes'], array( 'special-page') );
			register_block_pattern( $pattern['name'], $pattern );
		}
	}
```


### I want to totally remove this functionality from my "Special Post" post type.

```
	$patterns = WP_Block_Patterns_Registry::get_instance()->get_all_registered();
	foreach ( $patterns as $pattern ) {
		if (
			! empty( $pattern['startContentPostTypes'] ) &&
			in_array( 'special-post', $pattern['postTypes'] )
		) {
			unregister_block_pattern( $pattern['name'] );
			$pattern['postTypes'] = array_diff( $pattern['postTypes'], array( 'special-post') );
			register_block_pattern( $pattern['name'], $pattern );
		}
	}
```

## Testing Instructions
I applied the following diff:
```
diff --git a/lib/compat/wordpress-6.0/block-patterns-update.php b/lib/compat/wordpress-6.0/block-patterns-update.php
index c17a65be2a..fca9c8326b 100644
--- a/lib/compat/wordpress-6.0/block-patterns-update.php
+++ b/lib/compat/wordpress-6.0/block-patterns-update.php
@@ -23,7 +23,8 @@ function gutenberg_register_gutenberg_patterns() {
 	$patterns = array(
 		'query-standard-posts'                 => array(
 			'title'      => _x( 'Standard', 'Block pattern title', 'gutenberg' ),
-			'blockTypes' => array( 'core/query' ),
+			'blockTypes' => array( 'core/query', 'core/post-content' ),
+			'postTypes' => array( 'page', 'post' ),
 			'categories' => array( 'query' ),
 			'content'    => '<!-- wp:query {"query":{"perPage":3,"pages":0,"offset":0,"postType":"post","order":"desc","orderBy":"date","author":"","search":"","exclude":[],"sticky":"","inherit":false}} -->
 							<div class="wp-block-query">
@@ -41,7 +42,7 @@ function gutenberg_register_gutenberg_patterns() {
 		),
 		'query-medium-posts'                   => array(
 			'title'      => _x( 'Image at left', 'Block pattern title', 'gutenberg' ),
-			'blockTypes' => array( 'core/query' ),
+			'blockTypes' => array( 'core/query', 'core/post-content' ),
 			'categories' => array( 'query' ),
 			'content'    => '<!-- wp:query {"query":{"perPage":3,"pages":0,"offset":0,"postType":"post","order":"desc","orderBy":"date","author":"","search":"","exclude":[],"sticky":"","inherit":false}} -->
 							<div class="wp-block-query">
@@ -61,7 +62,8 @@ function gutenberg_register_gutenberg_patterns() {
 		),
 		'query-small-posts'                    => array(
 			'title'      => _x( 'Small image and title', 'Block pattern title', 'gutenberg' ),
-			'blockTypes' => array( 'core/query' ),
+			'blockTypes' => array( 'core/query', 'core/post-content' ),
+			'postTypes' => array( 'page', 'post' ),
 			'categories' => array( 'query' ),
 			'content'    => '<!-- wp:query {"query":{"perPage":3,"pages":0,"offset":0,"postType":"post","order":"desc","orderBy":"date","author":"","search":"","exclude":[],"sticky":"","inherit":false}} -->
 							<div class="wp-block-query">
@@ -80,7 +82,8 @@ function gutenberg_register_gutenberg_patterns() {
 		),
 		'query-grid-posts'                     => array(
 			'title'      => _x( 'Grid', 'Block pattern title', 'gutenberg' ),
-			'blockTypes' => array( 'core/query' ),
+			'blockTypes' => array( 'core/query','core/post-content' ),
+			'postTypes' => array( 'page' ),
 			'categories' => array( 'query' ),
 			'content'    => '<!-- wp:query {"query":{"perPage":6,"pages":0,"offset":0,"postType":"post","order":"desc","orderBy":"date","author":"","search":"","exclude":[],"sticky":"exclude","inherit":false},"displayLayout":{"type":"flex","columns":3}} -->
 							<div class="wp-block-query">
@@ -96,7 +99,8 @@ function gutenberg_register_gutenberg_patterns() {
 		),
 		'query-large-title-posts'              => array(
 			'title'      => _x( 'Large title', 'Block pattern title', 'gutenberg' ),
-			'blockTypes' => array( 'core/query' ),
+			'blockTypes' => array( 'core/query', 'core/post-content' ),
+			'postTypes' => array( 'post' ),
 			'categories' => array( 'query' ),
 			'content'    => '<!-- wp:group {"align":"full","style":{"spacing":{"padding":{"top":"100px","right":"100px","bottom":"100px","left":"100px"}},"color":{"text":"#ffffff","background":"#000000"}}} -->
 							<div class="wp-block-group alignfull has-text-color has-background" style="background-color:#000000;color:#ffffff;padding-top:100px;padding-right:100px;padding-bottom:100px;padding-left:100px"><!-- wp:query {"query":{"perPage":3,"pages":0,"offset":0,"postType":"post","order":"desc","orderBy":"date","author":"","search":"","exclude":[],"sticky":"","inherit":false}} -->

```

I created a new post and verified the following modal appears:
![image](https://user-images.githubusercontent.com/11271197/174779493-0d365d11-ccab-4652-b3cc-1ee4e9082e06.png)


I created a new page and verified the following modal appears:
![image](https://user-images.githubusercontent.com/11271197/174779554-b871d6d7-16d8-4661-af5d-19df8b27c33a.png)

jorgefilipecosta committed:
Add: Ability to use creation patterns for other post types besides page

mtias commented on packages/edit-post/src/components/start-page-options/index.js:
```diff
@@ -19,11 +19,20 @@ import { store as editPostStore } from '../../store';
+				: []
+			).concat(
+				__experimentalGetPatternsByBlockTypes(
+					`core/post-content/${ postType }`
```
This is a very weird construct to me since it seems to imply the post type is a child of the block type. Why can't we go with a separate key for post types? Seems both clearer and more flexible.

ntsekouras commented on lib/compat/wordpress-6.1/class-gutenberg-rest-block-patterns-controller.php:
```diff
@@ -0,0 +1,224 @@
+ *
+ * @see WP_REST_Controller
+ */
+class Gutenberg_REST_Block_Patterns_Controller extends WP_REST_Controller {
```
We need to align with the existing controller in core. This means, I believe we need to move forwards this issue: https://github.com/WordPress/gutenberg/pull/40902 and see what we'll need to update. Perhaps something like `WP_Theme_JSON_6_1 extends WP_Theme_JSON_6_0` might be needed.. 

Noting that I didn't check the code here - just some thoughts..

jorgefilipecosta replied:
Hi @ntsekouras In this PR I'm already aligining the endpoint with core (by accident) private $remote_patterns_loaded; is part of this PR. Basically, I copied this class from core, changed its name, and did the changes I needed. I'm using WordPress filters to overwrite the core class with this new Gutenberg one Following this approach we don't need any extends etc. We are free to apply the changes we need. When backporting the class to core we just copy this file back to the core, change its name, and that's it. I left a comment on @anton-vlasenko PR https://github.com/WordPress/gutenberg/pull/40902#pullrequestreview-1013387744.

ntsekouras reviewed (commented):
Thanks for exploring this Jorge!

I think that we should have an API that is not so specific to create `page` patterns, but instead we should explore a way for `postTypes` in general. What we might also need is a way to show conditionally patterns based on a template or template type(`archive, etc..`). So it comes to having an API that will handle which patterns are shown based on `context`.

I'm not saying right now that we should combine these(it would need exploration), but is something to have in mind.

We shouldn't have the example of `'core/template-part/header'` in mind because it's about block types and block variations.

Whatever API we will introduce, I believe it should be a new selector that will handle all the restrictions we impose and have a single source of truth. For example block patterns by `block type` are shown/requested in specific places to suggest a pattern and these patterns should also be filtered by the `postType`. Makes sense?

mtias commented:
@ntsekouras @jorgefilipecosta to connect the loop, I proposed a different API for this here: https://github.com/WordPress/gutenberg/issues/41398

ntsekouras commented:
>@ntsekouras @jorgefilipecosta to connect the loop, I proposed a different API for this here: https://github.com/WordPress/gutenberg/issues/41398

Yes, that was what I was suggesting 😄

jorgefilipecosta commented:
Hi @mtias, @ntsekouras  this PR passed by an update to the API.
I introduced a startContentPostTypes property in patterns that allows patterns to explicitly specify the post types where the pattern can be used in the start post modal.
I'm not using a postType property because it seems according to https://github.com/WordPress/gutenberg/issues/41398 that the idea of postType property is to restric the postTypes where the pattern can be used. I can implement that property in a parallel PR.
I may have a pattern that I want to make available in all post types but should only be available in the start content modal of Product. Although that pattern is ideal start for a product post it may make sense in certain cases to insert on a page.
Being able to use a pattern in a post type at all and wanting the pattern to appear in the start modal of a post type seem to be different concepts so I think using separate properties opens more possibilities.

But I can change the property to postTypes if we prefer to unify the concepts and not have two properties. The change is very easy I just need to replace startContentPostTypes with postTypes. The downside is that if a pattern is the start content of a postType I will not be able to use it at all in other postTypes.
Let me know your thoughts and thank you in advance for the feedback.

jorgefilipecosta committed:
linting

mcsf commented:
Hey @jorgefilipecosta, thanks for presenting these different alternatives.

I vote for just introducing `postTypes`, per #41398.

I'd rather not rush to commit to a starter-content-related API, and instead make sure we offer simple APIs with clear semantics (like the `postType` field) and reliable pluggable systems (a proper `'register_block_patterns'` server-side action, per #40305, to let admins and plugins make their own rules about which patterns belong where).

jorgefilipecosta commented:
> Hey @jorgefilipecosta, thanks for presenting these different alternatives.
> 
> I vote for just introducing `postTypes`, per #41398.
> 
> I'd rather not rush to commit to a starter-content-related API, and instead make sure we offer simple APIs with clear semantics (like the `postType` field) and reliable pluggable systems (a proper `'register_block_patterns'` server-side action, per #40305, to let admins and plugins make their own rules about which patterns belong where).

Hi @mcsf, so just to make sure we are in agreement, in that case, it is a fair tradeoff and ok that if a pattern is supposed to appear in the start content of a "product" post type it will not be available on any other post type at all?
Page post type is a special case, where by default if a pattern declares "core/post-content" it appears as the start content of pages, and it is still possible to use the pattern in other post types (exactly what already happens now on 6.0 so we are being back-compatible), so the page is the only post type where a pattern can be the start content and also be used in other post types. If one wants to really restrict to pages it is possible to do "postTypes": array('page') like for any other post types.

mtias commented:
> it is a fair tradeoff and ok that if a pattern is supposed to appear in the start content of a "product" post type it will not be available on any other post type at all?

Correct. We might want in the future to allow post types to opt-in to starter modal wholesale, like "page" is auto-opted in, but it might also be entirely unnecessary if it's predicated on the patterns. From a pattern perspective, when we say it is restricted to `core/post-content` we are saying it belongs as direct child of post content. Right now we interpret that as showing up in the starter modal, but also means showing up in the patterns inserter for the root / post-content area. The `postType` is a parallel property that furthers controls the experience. You should be able to declare patterns that only show up in the post type, and patterns that only show up in that post type + use the start modal. That's why combining postTypes and blockTypes properties is useful.

jorgefilipecosta committed:
refactor to postTypes property

jorgefilipecosta commented:
Thank you @mtias, @mcsf, @ntsekouras  this PR was updated following your suggestions and now we have a postTypes property. I'm now implementing as a follow-up the postType restriction part (if a postType is product the pattern will not appear at all for other post types).

jorgefilipecosta commented:
A follow-up PR was proposed at https://github.com/WordPress/gutenberg/pull/41842 that implements the postTypes restriction (if a pattern specifies postTypes it will only appear on the postTypes it specifies).
I think both PRs are ready for a review and can be merged.

jorgefilipecosta commented on lib/compat/wordpress-6.1/class-gutenberg-rest-block-patterns-controller.php:
```diff
@@ -0,0 +1,224 @@
+<?php
```
*Note to reviewers:* this file is an exact copy of core with the following changes:
- class name changed for the Gutenberg prefix.
- post_types added to the description in public function get_item_schema().
- 'postTypes'     => 'post_types', added to $key in prepare_item_for_response.

There are no other changes to the core file.

jorgefilipecosta committed:
fix merge issue.

mcsf committed:
Compat: Pull _register_theme_block_patterns into 6.1, replacing 6.0 shim

No changes from Core's _register_theme_block_patterns, just a copy in
preparation for the next commit.

mcsf committed:
gutenberg_register_theme_block_patterns: support postTypes field

jorgefilipecosta commented:
Thank you @mcsf, for including support to the post types in the theme HTML-defined patterns. I think this PR is ready to land and just needs the green approval.

mtias commented on packages/edit-post/src/components/start-page-options/index.js:
```diff
@@ -17,15 +17,30 @@ import { store as editorStore } from '@wordpress/editor';
+	return useMemo( () => {
+		return blockPatterns.filter( ( pattern ) => {
+			return (
+				( postType === 'page' && ! pattern.postTypes ) ||
```
Can we include some comments here? The logic and its implications are worth stating since it's not obvious on its own.

jorgefilipecosta replied:
I included some inline comments.

Mamaduka replied:
Sorry, I'm a bit late. But what do you think if we make filtering by post type part of REST API request? There's no need to send a full response if we filter out results on the client-side.

P.S. We're already doing something similar for the templates - https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php#L747-L750

mtias commented on packages/core-data/src/resolvers.js:
```diff
@@ -480,6 +480,8 @@ export const getBlockPatterns =
 				switch ( key ) {
 					case 'block_types':
 						return 'blockTypes';
+					case 'post_types':
```
Is there any documentation to update around this? cc @juanmaguitar

mcsf replied:
Can we generalise the casing transformation?

```js
// hello_world -> helloWorld
function snakeToCamel( string ) {
  return string.replace( /_([a-z])/, ( _, letter ) => letter.toUpperCase() );
}

mapKeys( pattern, ( _, phpKey ) => snakeToCamel( phpKey ) );
```

If not (why not?), then it would be nice to eventually be able to source the [mapping from Core](https://github.com/WordPress/wordpress-develop/blob/e7cafe6141631bc8125c3173404edab09c868961/src/wp-includes/block-patterns.php#L201-L210) rather than keep a copy here.

jorgefilipecosta replied:
I applied the feedback, using a function from lodash that is already used in other parts of the codebase.

mcsf commented on lib/compat/wordpress-6.1/class-gutenberg-rest-block-patterns-controller.php:
```diff
@@ -0,0 +1,224 @@
+					'readonly'    => true,
+					'context'     => array( 'view', 'edit', 'embed' ),
+				),
+				'post_types'     => array(
+					'description' => __( 'Post types where the pattern is intended to be used as the starting content.' ),
+					'type'        => 'array',
+					'readonly'    => true,
+					'context'     => array( 'view', 'edit', 'embed' ),
+				),
```
I am not sure what the best way to track this is, but we will also need to update the documentation of Core's `register_block_pattern`:

https://github.com/WordPress/wordpress-develop/blob/e7cafe6141631bc8125c3173404edab09c868961/src/wp-includes/class-wp-block-patterns-registry.php#L47-L72

jorgefilipecosta replied:
I added a PR against core here https://github.com/WordPress/wordpress-develop/pull/2858 so we don't miss the required change in core.

mcsf reviewed (commented):
Looking good, just some little bits left.

jorgefilipecosta committed:
Documentation added and feedback applied.

jorgefilipecosta committed:
Improve documentation

jorgefilipecosta commented:
All the feedback was applied and some documentation was included.

mcsf reviewed (approved):
🚢

jorgefilipecosta committed:
lint fixes

bph commented:
[DevNote for Misc Block Editor Post](https://github.com/WordPress/gutenberg/issues/42961#issuecomment-1275081692):
